### PR TITLE
Searcher: support precise lang filters

### PR DESCRIPTION
**Pull request by jtibshirani**
This PR adds support for the `search-content-based-lang-detection` feature flag
to unindexed search. It introduces a new type of matcher called `langMatcher`
that uses go-enry to determine a file's language.

The PR is big, but it's broken into commits. Main pieces:
* Refactor `search_regex` to introduce a `fileLoader` object
* Add language matching using go-enry
* Introduce new `IncludeLangs` and `ExcludeLangs` fields so we can pipe
language filters directly to searcher

Most file changes are just due renaming `IncludePatterns` -> `IncludePaths`.

## Test plan

Added unit tests and benchmarks

### CODE REVIEW

**Comment by jtibshirani** [comment]
Here are the searcher benchmarks before vs. after the change. I grouped each benchmark type together. Significant differences noted with `*...*`.

<details>
  <summary>Benchmark comparison </summary>

```
BenchmarkSearchRegex_large_fixed-10                  	     219	   5372972 ns/op	15692829 B/op	     243 allocs/op
BenchmarkSearchRegex_large_fixed-10                  	     216	   5427758 ns/op	15700592 B/op	     244 allocs/op

BenchmarkSearchRegex_rare_fixed-10                   	     168	   7246439 ns/op	15693014 B/op	     657 allocs/op
BenchmarkSearchRegex_rare_fixed-10                   	     169	   7591177 ns/op	15637489 B/op	     661 allocs/op

BenchmarkSearchRegex_large_fixed_casesensitive-10    	     436	   2765823 ns/op	   14090 B/op	     194 allocs/op
BenchmarkSearchRegex_large_fixed_casesensitive-10    	     441	   2716651 ns/op	   14797 B/op	     195 allocs/op

BenchmarkSearchRegex_large_empty_pattern-10    	     231	   5122012 ns/op	16780675 B/op	     105 allocs/op
BenchmarkSearchRegex_large_empty_pattern-10          	     228	   5145705 ns/op	16786850 B/op	     107 allocs/op

*Before: N/A*
*BenchmarkSearchRegex_large_lang_filter_common-10    	     261	   4557803 ns/op	15655989 B/op	     320 allocs/op*

*Before: N/A*
*BenchmarkSearchRegex_large_lang_filter_uncommon-10    	     974	   1274682 ns/op	15612261 B/op	     348 allocs/op*

BenchmarkSearchRegex_large_re_dotstar-10             	       5	 259080342 ns/op	628672656 B/op	 6098823 allocs/op
BenchmarkSearchRegex_large_re_dotstar-10             	       5	 224910450 ns/op	628902502 B/op	 6098810 allocs/op

BenchmarkSearchRegex_large_re_common-10              	     199	   6241995 ns/op	 7503550 B/op	   68778 allocs/op
BenchmarkSearchRegex_large_re_common-10              	     222	   5202429 ns/op	 7500757 B/op	   68776 allocs/op

BenchmarkSearchRegex_large_re_anchor-10              	      19	  57122230 ns/op	 4921381 B/op	   50146 allocs/op
BenchmarkSearchRegex_large_re_anchor-10              	      20	  57729229 ns/op	 4934114 B/op	   50150 allocs/op

BenchmarkSearchRegex_large_capture_group-10          	       8	 137275490 ns/op	 3449010 B/op	   25261 allocs/op
BenchmarkSearchRegex_large_capture_group-10          	       7	 147519845 ns/op	 3465006 B/op	   25264 allocs/op

*BenchmarkSearchRegex_large_path/path_only-10         	    7681	    152459 ns/op	    5266 B/op	      72 allocs/op*
*BenchmarkSearchRegex_large_path/path_only-10         	    1627	    681527 ns/op	    8269 B/op	     113 allocs/op*

BenchmarkSearchRegex_large_path/content_only-10      	     171	   6916182 ns/op	   64785 B/op	     652 allocs/op
BenchmarkSearchRegex_large_path/content_only-10      	     172	   7277534 ns/op	   66118 B/op	     653 allocs/op

BenchmarkSearchRegex_large_path/both_path_and_content-10         	     169	   7093483 ns/op	   65127 B/op	     652 allocs/op
BenchmarkSearchRegex_large_path/both_path_and_content-10         	     172	   6946727 ns/op	   63743 B/op	     631 allocs/op

BenchmarkSearchRegex_small_fixed-10                              	    7720	    156705 ns/op	  309957 B/op	     128 allocs/op
BenchmarkSearchRegex_small_fixed-10                              	    8428	    142722 ns/op	  324529 B/op	     130 allocs/op

BenchmarkSearchRegex_small_fixed_casesensitive-10                	   12189	    100133 ns/op	    9670 B/op	     122 allocs/op
BenchmarkSearchRegex_small_fixed_casesensitive-10                	   12505	     95307 ns/op	   10277 B/op	     123 allocs/op

BenchmarkSearchRegex_small_re_dotstar-10                         	     436	   2730551 ns/op	 4108227 B/op	   40140 allocs/op
BenchmarkSearchRegex_small_re_dotstar-10                         	     517	   2273420 ns/op	 4126375 B/op	   40142 allocs/op

BenchmarkSearchRegex_small_re_common-10                          	   10000	    108978 ns/op	   48431 B/op	     530 allocs/op
BenchmarkSearchRegex_small_re_common-10                          	   10000	    101683 ns/op	   49068 B/op	     531 allocs/op

BenchmarkSearchRegex_small_re_anchor-10                          	    1408	    840718 ns/op	   25483 B/op	     334 allocs/op
BenchmarkSearchRegex_small_re_anchor-10                          	    1626	    726375 ns/op	   26301 B/op	     335 allocs/op

BenchmarkSearchRegex_small_capture_group-10                      	     684	   1816197 ns/op	   24139 B/op	     315 allocs/op
BenchmarkSearchRegex_small_capture_group-10                      	     759	   1574914 ns/op	   24780 B/op	     316 allocs/op
```

Notes:
* The language filtering overhead isn't significant in common case where languages have unambiguous extensions
* Path-only searches became slightly slower because I needed to remove a dedicated "fast path" optimization

File: cmd/searcher/internal/search/langmatch.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+type enryLangMatcher struct {
+	IncludeLangs []string
+	ExcludeLangs []string
+	OnlyExcludes bool
+}
+
+func (em *enryLangMatcher) Matches(path string, getContent func() ([]byte, error)) bool {
+	// We use Sourcegraph's wrapper around enry because it supports lazily fetching
+	// content and contains some optimizations for ambiguous extensions.
+	langs, err := languages.GetLanguages(path, getContent)
```
Comment (jtibshirani) [inline cmd/searcher/internal/search/langmatch.go:36]: We don't truncate the content or anything here -- I didn't see an issue in benchmarks, and we already enforce a max file size of 2MB.

File: cmd/searcher/internal/search/langmatch.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+	// It's fine if file has no detected language, as long as there are no include filters
+	if len(langs) == 0 {
+		return em.OnlyExcludes
+	}
+
+	// Choose the most likely language
+	lang := langs[0]
+	for _, includeLang := range em.IncludeLangs {
+		if lang != includeLang {
```
Comment (jtibshirani) [inline cmd/searcher/internal/search/langmatch.go:51]: These semantics are pretty weird -- you need to match *all* of the included languages. Since each document has one language, it's never possible to match more than one include filter.

So queries like `lang:Go lang:Markdown` always return zero results. I wonder if we should just reject queries with multiple language filters up-front.

File: cmd/searcher/protocol/searcher.go [hunk 3]
```diff
@@ -113,6 +114,10 @@ type PatternInfo struct {
 	// use it since selection is done after the query completes, but exposing it can enable
 	// optimizations.
 	Select string
+
+	// Languages represents the set of languages requested in the query. It is only uesd for
+	// structural search and is separate from IncludeLangs, which represents language filters.
+	Languages []string
```
Comment (jtibshirani) [inline cmd/searcher/protocol/searcher.go:120]: We can remove this in favor of `IncludeLangs` if the feature flag ever becomes the default, since we'll be sure `IncludeLangs` is always populated.

**Comment by keegancsmith** [comment]
> Here are the searcher benchmarks before vs. after the change. I grouped each benchmark type together. Significant differences noted with `*...*`.

I encourage the use of the benchstat command, much easier to parse and find changes. You need to run your benchmarks with `-count` to collect a few samples so the statistical comparison is fair. But here are the results on my desktop machine:

```
name                                             old time/op    new time/op    delta
SearchRegex_large_fixed-32                         5.93ms ±10%    6.32ms ± 8%      ~     (p=0.310 n=5+5)
SearchRegex_rare_fixed-32                          5.65ms ± 9%    5.79ms ±10%      ~     (p=0.690 n=5+5)
SearchRegex_large_fixed_casesensitive-32           3.09ms ± 3%    3.19ms ± 1%      ~     (p=0.095 n=5+5)
SearchRegex_large_re_dotstar-32                     259ms ± 2%     258ms ± 3%      ~     (p=1.000 n=5+5)
SearchRegex_large_re_common-32                     7.27ms ± 8%    7.16ms ± 9%      ~     (p=0.548 n=5+5)
SearchRegex_large_re_anchor-32                     66.9ms ± 1%    67.3ms ± 2%      ~     (p=0.690 n=5+5)
SearchRegex_large_capture_group-32                  174ms ± 5%     181ms ±10%      ~     (p=0.310 n=5+5)
SearchRegex_large_path/path_only-32                 153µs ± 1%     390µs ± 3%  +155.04%  (p=0.008 n=5+5)
SearchRegex_large_path/content_only-32             12.7ms ±10%    11.5ms ± 8%    -9.53%  (p=0.032 n=5+5)
SearchRegex_large_path/both_path_and_content-32    11.4ms ±22%    11.2ms ± 8%      ~     (p=0.841 n=5+5)
SearchRegex_small_fixed-32                          147µs ± 1%     149µs ± 1%    +1.77%  (p=0.008 n=5+5)
SearchRegex_small_fixed_casesensitive-32           95.5µs ± 3%    92.4µs ±10%      ~     (p=0.310 n=5+5)
SearchRegex_small_re_dotstar-32                    4.34ms ± 1%    4.34ms ± 2%      ~     (p=0.841 n=5+5)
SearchRegex_small_re_common-32                      170µs ± 2%     171µs ± 1%      ~     (p=0.690 n=5+5)
SearchRegex_small_re_anchor-32                     1.13ms ±35%    1.49ms ± 5%      ~     (p=0.222 n=5+5)
SearchRegex_small_capture_group-32                 3.41ms ±10%    3.62ms ± 7%      ~     (p=0.151 n=5+5)

name                                             old alloc/op   new alloc/op   delta
SearchRegex_large_fixed-32                         15.7MB ± 0%    15.7MB ± 0%      ~     (p=0.841 n=5+5)
SearchRegex_rare_fixed-32                          15.7MB ± 0%    15.7MB ± 0%      ~     (p=0.151 n=5+5)
SearchRegex_large_fixed_casesensitive-32           14.4kB ± 4%    15.4kB ± 2%    +6.90%  (p=0.008 n=5+5)
SearchRegex_large_re_dotstar-32                     630MB ± 0%     630MB ± 0%      ~     (p=0.548 n=5+5)
SearchRegex_large_re_common-32                     7.70MB ± 0%    7.70MB ± 0%      ~     (p=0.841 n=5+5)
SearchRegex_large_re_anchor-32                     4.95MB ± 0%    4.95MB ± 1%      ~     (p=0.690 n=5+5)
SearchRegex_large_capture_group-32                 3.46MB ± 1%    3.45MB ± 1%      ~     (p=0.690 n=5+5)
SearchRegex_large_path/path_only-32                5.27kB ± 0%    9.02kB ± 1%   +71.02%  (p=0.008 n=5+5)
SearchRegex_large_path/content_only-32             67.3kB ± 4%    68.2kB ± 4%      ~     (p=0.841 n=5+5)
SearchRegex_large_path/both_path_and_content-32    66.7kB ± 3%    66.8kB ± 4%      ~     (p=1.000 n=5+5)
SearchRegex_small_fixed-32                          473kB ± 0%     474kB ± 0%    +0.23%  (p=0.008 n=5+5)
SearchRegex_small_fixed_casesensitive-32           9.90kB ± 0%   11.04kB ± 0%   +11.52%  (p=0.008 n=5+5)
SearchRegex_small_re_dotstar-32                    4.17MB ± 0%    4.17MB ± 0%      ~     (p=0.095 n=5+5)
SearchRegex_small_re_common-32                     50.2kB ± 0%    51.4kB ± 0%    +2.40%  (p=0.008 n=5+5)
SearchRegex_small_re_anchor-32                     25.8kB ± 1%    27.0kB ± 2%    +4.71%  (p=0.008 n=5+5)
SearchRegex_small_capture_group-32                 24.5kB ± 2%    25.8kB ± 2%    +5.20%  (p=0.008 n=5+5)

name                                             old allocs/op  new allocs/op  delta
SearchRegex_large_fixed-32                            237 ± 1%       254 ± 1%    +7.25%  (p=0.008 n=5+5)
SearchRegex_rare_fixed-32                             674 ± 0%       692 ± 0%    +2.67%  (p=0.016 n=5+4)
SearchRegex_large_fixed_casesensitive-32              194 ± 0%       211 ± 0%    +8.76%  (p=0.008 n=5+5)
SearchRegex_large_re_dotstar-32                     6.10M ± 0%     6.10M ± 0%      ~     (p=0.056 n=5+5)
SearchRegex_large_re_common-32                      68.9k ± 0%     68.9k ± 0%    +0.02%  (p=0.008 n=5+5)
SearchRegex_large_re_anchor-32                      50.2k ± 0%     50.2k ± 0%    +0.03%  (p=0.008 n=5+5)
SearchRegex_large_capture_group-32                  25.3k ± 0%     25.3k ± 0%      ~     (p=0.095 n=5+5)
SearchRegex_large_path/path_only-32                  72.0 ± 0%     129.0 ± 0%   +79.17%  (p=0.008 n=5+5)
SearchRegex_large_path/content_only-32                653 ± 0%       669 ± 0%    +2.54%  (p=0.008 n=5+5)
SearchRegex_large_path/both_path_and_content-32       653 ± 0%       647 ± 0%    -0.80%  (p=0.008 n=5+5)
SearchRegex_small_fixed-32                            136 ± 0%       153 ± 0%   +12.50%  (p=0.016 n=4+5)
SearchRegex_small_fixed_casesensitive-32              122 ± 0%       139 ± 0%   +13.93%  (p=0.008 n=5+5)
SearchRegex_small_re_dotstar-32                     40.1k ± 0%     40.2k ± 0%    +0.04%  (p=0.016 n=4+5)
SearchRegex_small_re_common-32                        530 ± 0%       547 ± 0%    +3.21%  (p=0.008 n=5+5)
SearchRegex_small_re_anchor-32                        334 ± 0%       351 ± 0%    +5.09%  (p=0.008 n=5+5)
SearchRegex_small_capture_group-32                    315 ± 0%       332 ± 0%    +5.40%  (p=0.008 n=5+5)
```

It seems the main differences the benchmarks are noticing are all around allocations/memory usage. However, the significant changes here are all on something which in real terms is small amounts of memory. So seems acceptable.

File: cmd/searcher/internal/search/langmatch.go (hunk 1 above)
Comment (keegancsmith) [reply cmd/searcher/internal/search/langmatch.go:36]: I wonder if this API is the source of the increase in allocations? I'd be all for making it take in a langs buffer for it to use to avoid allocating. Then we just need to ensure we don't use the lang matcher concurrently, which should be doable since we already take care for that in the regexp matchers via the clone methods?

File: cmd/searcher/internal/search/search_regex.go [hunk 4]
```diff
@@ -193,6 +190,48 @@ func regexSearch(
...
+	// re-allocating. It is only used if we need to transform the input
+	// before matching. For example we lower case the input in the case of
+	// ignoreCase.
+	scratchBuf []byte
+}
+
+func (l *fileLoader) load(f *srcFile) {
+	if f == l.prevFile {
+		return
+	}
```
Comment (keegancsmith) [inline cmd/searcher/internal/search/search_regex.go:217]: looks like you forget to set prevFile?

**Review by keegancsmith — APPROVED** [review]
LGTM, see inline feedback.

> Most file changes are just due renaming IncludePatterns -> IncludePaths.

Why did you do the rename?

File: cmd/searcher/internal/search/search_regex.go [hunk 5]
```diff
@@ -122,54 +109,64 @@ func regexSearch(
...
+				// Check if the pattern matches
+				if patternMatchesPaths {
 					match = m.MatchesString(f.Name)
-					if match {
-						fm.Path = f.Name
+				}
+
+				if !match && patternMatchesContent {
+					if _, ok := m.(allMatchTree); ok {
+						// Avoid loading the file is this pattern always matches
```
Comment (varungandhi-src) [inline cmd/searcher/internal/search/search_regex.go:158]:
```suggestion
						// Avoid loading the file if this pattern always matches
```

Typo

File: cmd/searcher/internal/search/search_regex.go [hunk 6]
```diff
@@ -56,6 +57,7 @@ func regexSearch(
 	ctx context.Context,
 	m matchTree,
 	pm *pathMatcher,
+	lm langMatcher,
```
Comment (varungandhi-src) [inline cmd/searcher/internal/search/search_regex.go:60]: Maybe there is a strong convention in the code around short variable names here, but keeping track of all the 1-2 letter variable names here (I see `l`, `f`, `fm`, `m`, `pm`, `lm`, `zf`) seems a bit hard to follow.

File: cmd/searcher/internal/search/search_regex.go [hunk 7]
```diff
@@ -193,6 +190,48 @@ func regexSearch(
...
+// file it loaded, in case it's requested twice (for example first from language
+// detection, then pattern matching).
+type fileLoader struct {
+	zf              *zipFile
+	isCaseSensitive bool
+
+	// fileMatchBuf is what we run match on, fileBuf is the original data
+	// (used for the content preview).
+	fileBuf      []byte
+	fileMatchBuf []byte
```
Comment (varungandhi-src) [inline cmd/searcher/internal/search/search_regex.go:203]:
```suggestion
    // fullContentsBuf is the original data for the file, and is used for the content preview
	fullContentsBuf      []byte
	// matchBuf represents a (potentially case converted) sub-slice of fullContentsBuf
	// used for matching content or languages
	matchBuf []byte
```

Minor suggestion for renaming to reduce similarity between the field names.

(That said, I'm not entirely sure where the 2MB limit is being applied; is `fullContentsBuf` guaranteed to be less than 2MB?)

File: cmd/searcher/internal/search/langmatch.go (hunk 1 above)
Comment (varungandhi-src) [reply cmd/searcher/internal/search/langmatch.go:36]: > I didn't see an issue in benchmarks

Out of curiosity, which benchmarks did you run?

**Review by varungandhi-src — COMMENTED** [review]
Left some minor comments as I stumbled into this PR from the over-arching issue; nothing major as I'm not too familiar with this code.

File: cmd/searcher/internal/search/search_regex.go (hunk 4 above)
Comment (jtibshirani) [reply cmd/searcher/internal/search/search_regex.go:217]: 🤦 I'll fix this and rerun benchmarks. Also points to the need for a unit test for `fileLoader`.

File: internal/searcher/v1/searcher.proto [hunk 8]
```diff
@@ -216,4 +216,8 @@ message PatternInfo {
   // query defines the query to match files against. It supports regexp
   // patterns optionally combined through boolean operators.
   QueryNode query = 16;
+
+  // include_langs and exclude_langs represent the languages to filter on
+  repeated string include_langs = 17;
+  repeated string exclude_langs = 18;
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply internal/searcher/v1/searcher.proto:222]: By the time we pass queries to the backend services, both file filters and lang filters have been pulled out and only apply at the top level. They cannot be nested within other query structures. So our `QueryNode` only represents the "pattern", consisting of literal values, regexes, and boolean combinations of them.

File: cmd/searcher/internal/search/search_regex.go (hunk 7 above)
Comment (jtibshirani) [reply cmd/searcher/internal/search/search_regex.go:203]: Thanks for the suggestion here -- I improved the comments. For names, I'm actually going to keep the original since they better match what the variables represent: `fileMatchBuf` is not a subslice of `fileBuf`, it is either exactly the same, or a lower-cased version of it.

File: cmd/searcher/internal/search/langmatch.go (hunk 1 above)
Comment (jtibshirani) [reply cmd/searcher/internal/search/langmatch.go:36]: Could you clarify this comment? I believe this API just accepts `fileMatchBuf`, and it operates on that directly and doesn't reallocate. Or do you mean its return result, the `[]string` of detected languages?

**Comment by jtibshirani** [comment]
> I encourage the use of the benchstat command, 

Thank you, I wanted something like this but totally forgot it existed.

> Why did you do the rename?

When working with this code, I felt `IncludePatterns` was ambiguous -- what is it matching on, the contents or file or langs? The new name `IncludePaths` clarifies what it filters on, and distinguishes it from `IncludeLangs`. Note to self: it would've been better to pull this rename into a separate PR since it made this harder to review.

**Comment by jtibshirani** [comment]
Thanks for the reviews! I tried to respond or fix everything. Happy to push any follow-ups if something's resolved.

Here are my final results from benchstat. They look good to me -- no significant regressions, and the new "lang filter" benchmarks have decent latency and allocs.

<details>

```
goos: darwin
goarch: arm64
pkg: github.com/sourcegraph/sourcegraph/cmd/searcher/internal/search
                                                │   old.txt    │               new.txt                │
                                                │    sec/op    │    sec/op     vs base                │
SearchRegex_large_fixed-10                        5.280m ± ∞ ¹   5.354m ± ∞ ¹         ~ (p=0.056 n=5)
SearchRegex_rare_fixed-10                         6.973m ± ∞ ¹   7.032m ± ∞ ¹    +0.85% (p=0.008 n=5)
SearchRegex_large_fixed_casesensitive-10          2.646m ± ∞ ¹   2.776m ± ∞ ¹    +4.94% (p=0.008 n=5)
SearchRegex_large_re_dotstar-10                   221.0m ± ∞ ¹   224.6m ± ∞ ¹         ~ (p=0.151 n=5)
SearchRegex_large_re_common-10                    5.203m ± ∞ ¹   5.345m ± ∞ ¹    +2.74% (p=0.008 n=5)
SearchRegex_large_re_anchor-10                    56.48m ± ∞ ¹   56.97m ± ∞ ¹    +0.86% (p=0.016 n=5)
SearchRegex_large_capture_group-10                139.5m ± ∞ ¹   155.4m ± ∞ ¹   +11.39% (p=0.032 n=5)
SearchRegex_large_path/path_only-10               153.0µ ± ∞ ¹   752.9µ ± ∞ ¹  +392.00% (p=0.008 n=5)
SearchRegex_large_path/content_only-10            7.015m ± ∞ ¹   6.934m ± ∞ ¹    -1.16% (p=0.032 n=5)
SearchRegex_large_path/both_path_and_content-10   6.917m ± ∞ ¹   6.960m ± ∞ ¹         ~ (p=0.222 n=5)
SearchRegex_small_fixed-10                        141.1µ ± ∞ ¹   143.1µ ± ∞ ¹    +1.41% (p=0.008 n=5)
SearchRegex_small_fixed_casesensitive-10          98.11µ ± ∞ ¹   98.83µ ± ∞ ¹    +0.74% (p=0.008 n=5)
SearchRegex_small_re_dotstar-10                   2.267m ± ∞ ¹   2.256m ± ∞ ¹         ~ (p=0.222 n=5)
SearchRegex_small_re_common-10                    102.8µ ± ∞ ¹   102.8µ ± ∞ ¹         ~ (p=0.841 n=5)
SearchRegex_small_re_anchor-10                    725.8µ ± ∞ ¹   726.0µ ± ∞ ¹         ~ (p=1.000 n=5)
SearchRegex_small_capture_group-10                1.584m ± ∞ ¹   1.575m ± ∞ ¹         ~ (p=0.310 n=5)
SearchRegex_large_empty_pattern-10                               5.199m ± ∞ ¹
SearchRegex_large_lang_filter_common-10                          4.375m ± ∞ ¹
SearchRegex_large_lang_filter_uncommon-10                        1.213m ± ∞ ¹
geomean                                           2.895m         3.209m         +12.11%
¹ need >= 6 samples for confidence interval at level 0.95

                                                │    old.txt    │                new.txt                │
                                                │     B/op      │      B/op       vs base               │
SearchRegex_large_fixed-10                        14.97Mi ± ∞ ¹    14.97Mi ± ∞ ¹        ~ (p=0.841 n=5)
SearchRegex_rare_fixed-10                         14.97Mi ± ∞ ¹    14.97Mi ± ∞ ¹        ~ (p=0.095 n=5)
SearchRegex_large_fixed_casesensitive-10          13.76Ki ± ∞ ¹    14.84Ki ± ∞ ¹   +7.79% (p=0.008 n=5)
SearchRegex_large_re_dotstar-10                   599.8Mi ± ∞ ¹    599.7Mi ± ∞ ¹        ~ (p=0.841 n=5)
SearchRegex_large_re_common-10                    7.145Mi ± ∞ ¹    7.148Mi ± ∞ ¹        ~ (p=0.222 n=5)
SearchRegex_large_re_anchor-10                    4.692Mi ± ∞ ¹    4.695Mi ± ∞ ¹        ~ (p=1.000 n=5)
SearchRegex_large_capture_group-10                3.298Mi ± ∞ ¹    3.294Mi ± ∞ ¹        ~ (p=1.000 n=5)
SearchRegex_large_path/path_only-10               5.146Ki ± ∞ ¹    8.592Ki ± ∞ ¹  +66.94% (p=0.008 n=5)
SearchRegex_large_path/content_only-10            63.27Ki ± ∞ ¹    66.71Ki ± ∞ ¹   +5.44% (p=0.032 n=5)
SearchRegex_large_path/both_path_and_content-10   63.64Ki ± ∞ ¹    62.76Ki ± ∞ ¹   -1.38% (p=0.008 n=5)
SearchRegex_small_fixed-10                        321.9Ki ± ∞ ¹    346.1Ki ± ∞ ¹   +7.52% (p=0.008 n=5)
SearchRegex_small_fixed_casesensitive-10          9.505Ki ± ∞ ¹   10.592Ki ± ∞ ¹  +11.44% (p=0.008 n=5)
SearchRegex_small_re_dotstar-10                   3.934Mi ± ∞ ¹    3.934Mi ± ∞ ¹        ~ (p=0.421 n=5)
SearchRegex_small_re_common-10                    47.36Ki ± ∞ ¹    48.42Ki ± ∞ ¹   +2.24% (p=0.008 n=5)
SearchRegex_small_re_anchor-10                    25.01Ki ± ∞ ¹    26.15Ki ± ∞ ¹   +4.54% (p=0.008 n=5)
SearchRegex_small_capture_group-10                23.63Ki ± ∞ ¹    24.75Ki ± ∞ ¹   +4.75% (p=0.008 n=5)
SearchRegex_large_empty_pattern-10                                 16.01Mi ± ∞ ¹
SearchRegex_large_lang_filter_common-10                            14.93Mi ± ∞ ¹
SearchRegex_large_lang_filter_uncommon-10                          14.95Mi ± ∞ ¹
geomean                                           435.7Ki          805.0Ki         +5.92%
¹ need >= 6 samples for confidence interval at level 0.95

                                                │   old.txt    │               new.txt               │
                                                │  allocs/op   │  allocs/op    vs base               │
SearchRegex_large_fixed-10                         243.0 ± ∞ ¹    260.0 ± ∞ ¹   +7.00% (p=0.008 n=5)
SearchRegex_rare_fixed-10                          659.0 ± ∞ ¹    676.0 ± ∞ ¹   +2.58% (p=0.008 n=5)
SearchRegex_large_fixed_casesensitive-10           194.0 ± ∞ ¹    211.0 ± ∞ ¹   +8.76% (p=0.008 n=5)
SearchRegex_large_re_dotstar-10                   6.099M ± ∞ ¹   6.099M ± ∞ ¹        ~ (p=1.000 n=5)
SearchRegex_large_re_common-10                    68.77k ± ∞ ¹   68.79k ± ∞ ¹   +0.03% (p=0.008 n=5)
SearchRegex_large_re_anchor-10                    50.15k ± ∞ ¹   50.16k ± ∞ ¹   +0.03% (p=0.008 n=5)
SearchRegex_large_capture_group-10                25.27k ± ∞ ¹   25.28k ± ∞ ¹   +0.02% (p=0.008 n=5)
SearchRegex_large_path/path_only-10                72.00 ± ∞ ¹   129.00 ± ∞ ¹  +79.17% (p=0.008 n=5)
SearchRegex_large_path/content_only-10             652.0 ± ∞ ¹    669.0 ± ∞ ¹   +2.61% (p=0.008 n=5)
SearchRegex_large_path/both_path_and_content-10    652.0 ± ∞ ¹    647.0 ± ∞ ¹   -0.77% (p=0.008 n=5)
SearchRegex_small_fixed-10                         129.0 ± ∞ ¹    146.0 ± ∞ ¹  +13.18% (p=0.008 n=5)
SearchRegex_small_fixed_casesensitive-10           122.0 ± ∞ ¹    139.0 ± ∞ ¹  +13.93% (p=0.008 n=5)
SearchRegex_small_re_dotstar-10                   40.14k ± ∞ ¹   40.16k ± ∞ ¹   +0.04% (p=0.008 n=5)
SearchRegex_small_re_common-10                     530.0 ± ∞ ¹    547.0 ± ∞ ¹   +3.21% (p=0.008 n=5)
SearchRegex_small_re_anchor-10                     334.0 ± ∞ ¹    351.0 ± ∞ ¹   +5.09% (p=0.008 n=5)
SearchRegex_small_capture_group-10                 315.0 ± ∞ ¹    332.0 ± ∞ ¹   +5.40% (p=0.008 n=5)
SearchRegex_large_empty_pattern-10                                122.0 ± ∞ ¹
SearchRegex_large_lang_filter_common-10                           318.0 ± ∞ ¹
SearchRegex_large_lang_filter_uncommon-10                         348.0 ± ∞ ¹
geomean                                           1.844k         1.419k         +7.57%
¹ need >= 6 samples for confidence interval at level 0.95
```
</details>
